### PR TITLE
expose named export

### PR DESCRIPTION
**Pull request by akphi**
#### Checklist

I'm not as ambitious as #224, I'm just trying to make sure `named` exports work for `@fastify/cors`. I'm following what we already did in https://github.com/fastify/fastify

https://github.com/fastify/fastify/blob/acba25acc3e97e3598daf79537a39a5272dba7b8/fastify.js#L783-L796

```js
/**
 * These export configurations enable JS and TS developers
 * to consumer fastify in whatever way best suits their needs.
 * Some examples of supported import syntax includes:
 * - `const fastify = require('fastify')`
 * - `const { fastify } = require('fastify')`
 * - `import * as Fastify from 'fastify'`
 * - `import { fastify, TSC_definition } from 'fastify'`
 * - `import fastify from 'fastify'`
 * - `import fastify, { TSC_definition } from 'fastify'`
 */
module.exports = fastify
module.exports.fastify = fastify
module.exports.default = fastify
```

- [x] run `npm run test` and `npm run lint`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I dont want to close it without discussion but also dont want that somebody merges this without thinking the implications.


This plugin uses fastify-plugin. Fastify-plugin itself provides a named export. 

So you can do

`const { fastifyCors } = require('@fastify/cors')`

So i would not merge this, as we should already provide a named export.

**Comment by Uzlopak** [comment]
See https://github.com/fastify/fastify-plugin/blob/35a6288314ecdc610429c5675f01b4b453279796/plugin.js#L61

**Comment by kibertoad** [comment]
@Uzlopak we should at least adjust TS types to indicate that is possible then

**Comment by Uzlopak** [comment]
If you want to earn some brownie points than it would be good if we can fix the typings so that latest tsd passes and nodenext is supported. See the other two open PRs.

:)

**Comment by Uzlopak** [comment]
Ah wait. I thought @akphi answered...

But well. Still the same. We should fix nodenext

**Comment by akphi** [comment]
@Uzlopak what do you suppose we need for `nodenext` though? I'm not aware of the `js` part, so now you said it, I could omit my changes for `index.js` and just add the named export to `index.d.ts`.

**Comment by akphi** [comment]
@Uzlopak hm, compared to #230 I'm trying to make this work with the named import case for `NodeNext`, but you can see my repro here

https://github.com/akphi/issue-repo/pull/11

Even after patching the `index.d.ts` to support named export, i.e.

```ts
export { fastifyCors };
```

it doesn't work, I got the error like mentioned in that PR:

```
import { fastifyCors } from "@fastify/cors";
         ^^^^^^^^^^^
SyntaxError: Named export 'fastifyCors' not found. The requested module '@fastify/cors' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@fastify/cors';
const { fastifyCors } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12)
```

I believe I still need to make the changes I made in `index.js`

File: index.js [hunk 1]
```diff
@@ -220,7 +220,15 @@ function isRequestOriginAllowed (reqOrigin, allowedOrigin) {
   }
 }
 
-module.exports = fp(fastifyCors, {
+const _fastifyCors = fp(fastifyCors, {
```
Comment (Uzlopak) [inline index.js:223]: revert tiese changes, and I will approve and merge it ;)

**Review by Uzlopak — CHANGES_REQUESTED** [review]
Just a remakr

File: index.js (hunk 1 above)
Comment (akphi) [reply index.js:223]: This addresses the issue I found in https://github.com/akphi/issue-repo/pull/11 in my project using `NodeNext`. Is there something I miss here, but what you said about `fastify-plugin` just doesn't work for our Typescript project.

**Comment by Uzlopak** [comment]
Do we have an underlying issue in fastify-plugin?

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by Uzlopak** [comment]
@mcollina 

We should not merge this.

**Comment by Uzlopak** [comment]
I already did use Request Changes option.

We have already named export in fastify-plugin. So the question is, if it is actually properly working or not. I have the feeling that the issue is typescript and not Javascript.

**Comment by Eomm** [comment]
> We have already named export in fastify-plugin. 

Could you link it?

**Comment by Uzlopak** [comment]
Named Export, e.g. `const fastifyCors = require('@fastify/cors').fastifyCors`
https://github.com/fastify/fastify-plugin/blob/bdbe525a949e34f1fb9244e33b9d210ca0fe2b39/plugin.js#L58

.default Export, e.g. `const fastifyCors = require('@fastify/cors').default`
https://github.com/fastify/fastify-plugin/blob/bdbe525a949e34f1fb9244e33b9d210ca0fe2b39/plugin.js#L50

Default Export, e.g. `const fastifyCors = require('@fastify/cors')`
https://github.com/fastify/fastify-plugin/blob/bdbe525a949e34f1fb9244e33b9d210ca0fe2b39/plugin.js#L61

**Comment by akphi** [comment]
I tested with https://github.com/akphi/issue-repo/pull/11

```js
import { fastify } from "fastify";
import { fastifyCors } from "@fastify/cors";

const PORT = 9999;

const server = fastify({
  logger: true,
});

server.register(fastifyCors, {
  methods: ["OPTIONS"],
  origin: [/localhost/],
  credentials: true,
});

server.listen(
  {
    port: PORT,
  },
  (error, address) => {
    if (error) {
      throw error;
    }
  }
);
```

If you run the script `yarn start:js` you would get the error

```
import { fastifyCors } from "@fastify/cors";
         ^^^^^^^^^^^
SyntaxError: Named export 'fastifyCors' not found. The requested module '@fastify/cors' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@fastify/cors';
const { fastifyCors } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:533:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12)
```

I have tested by poking into `node_modules/@fastify/cors/index.js` and made the changes I made in this PR and things work fine. So this is not a problem just pertaining to TS.

**Comment by climba03003** [comment]
> So this is not a problem just pertaining to TS.

It is a TS problem, because TS will read the file and see if it suitable.
TS is not smart enough to see `fastify-plugin` syntax to be identical to `module.export.<name> = <func>`.

**Comment by akphi** [comment]
> > So this is not a problem just pertaining to TS.
> 
> It is a TS problem, because TS will read the file and see if it suitable. TS is not smart enough to see `fastify-plugin` syntax to be identical to `module.export.<name> = <func>`.
> 
> Of cause, the changes in `index.d.ts` is needed.

Sorry, I don't get it, when I run `node index.js` in that project, I don't get which part involves TS here, did I miss something?

**Comment by climba03003** [comment]
> > > So this is not a problem just pertaining to TS.
> > 
> > It is a TS problem, because TS will read the file and see if it suitable. TS is not smart enough to see `fastify-plugin` syntax to be identical to `module.export.<name> = <func>`.
> > 
> > Of cause, the changes in `index.d.ts` is needed.
> 
> But I faced the problem when I run `node index.js` in that project, I don't get which part involves TS here, did I miss something?

Strange, it certainly something changes in node. From reading the document, it's seems like the current import of commonjs will go through a static analysis and it only permit some of the syntax.

**Comment by Eomm** [comment]
Yes, something strange is happening here.
I replicated the issue and if you try:

```
const awaited = await import( "@fastify/cors");
console.log({awaited})

const { fastifyCors } = awaited.default;
console.log(fastifyCors);
```

YOu will see that there is an unwanted and additional `default` wrapper

image

**Comment by climba03003** [comment]
> Yes, something strange is happening here.

In the latest version of `node` document. `import default` now directly map to `module.exports` which means unless the static analysis think that it should be a `named` export. There is no way to provide `dynamic named export` like before.

> The ECMAScript Module Namespace representation of a CommonJS module is always a namespace with a default export key pointing to the CommonJS module.exports value.

**Comment by akphi** [comment]
@climba03003 does this justify my fix? Is there anything else you need?

**Review by Eomm — APPROVED** [review]

**Comment by Uzlopak** [comment]
Sry

**Review by climba03003 — APPROVED** [review]

**Comment by Uzlopak** [comment]
Imho we should investigate 
https://github.com/fastify/fastify-plugin/pull/200

Also I will have a look at the nodenext PR

**Comment by climba03003** [comment]
@Uzlopak 
From my understanding, it becomes unrelated to `TypeScript`. But how `node` itself try to resolve `CJS` from `ESM`.
I expect this kind of issue will be spread over the organization in the near future.

**Comment by kibertoad** [comment]
@climba03003 Don't we generally use "the infamous triplet" across organization, similarly how it was done in the original version of this PR, to address this issue?

**Comment by climba03003** [comment]
> Don't we generally use "the infamous triplet" across organization, similarly how it was done in the original version of this PR, to address this issue?

Yes, but the current `node` using static analysis when `CJS` is imported from `ESM`. Which means the current `fastify-plugin` no longer works for the "the infamous triplet".
We need to explicit use `module.exports` in all the plugin if we want to support both `CJS` and `ESM`. I don't know which version update change this behavior since `TypeScript` actually helps me to eliminate the `import` problem. But the truth is that only explicit `module.exports` will works.

**Comment by fox1t** [comment]
OK, the issue is now clear to me. Let me try to summarise it. When Node.js uses ESM context (so you are running an entry point with `.mjs` extension or have `"type": "module"` in your `package.json`) when you import a CJS package.json, Node.js will treat it as cjs regardless of the exports we provide and types we add, and it refuses to import a named export. I think that as we found the "famous triplet," we should try to find a superset of it to support also `"type": "module"` without losing all of the benefits that we already have. Not everyone is ready to go full ESM yet.

**Comment by climba03003** [comment]
> Node.js will treat it as cjs regardless of the exports we provide and types we add, and it refuses to import a named export.

It will support if you `explicit` specify in `module.exports`. 
Currently, we use `fastify-plugin` as dynamic generation of `named export` which is not supported by `node`.

```js
// supported in NodeJS and TypeScript
module.exports = plugin
module.exports.default = plugin
module.exports.plugin = plugin

// only supported in TypeScript
module.exports = fp(plugin)
```

**Comment by fox1t** [comment]
Yes! I've just played a little bit with a repro on my machine and we need to rely on that Node.js static analysis on CJS imports. Are we saying that we should "just" fix `fp` here?

**Comment by climba03003** [comment]
> Are we saying that we should "just" fix fp here?

@fox1t 

They are different problem. In here, we need to support native `nodejs`, so we need to use `module.exports`.
In `fastify-plugin`, it is the problem of how `TypeScript` resolving modules.

`TypeScript` resolve module based on `types` and `NodeJS` based on `static analysis` of `JavaScript` file.
Both patch are needed in their own aspect.

**Comment by fox1t** [comment]
Yes, I am aware of it. Probably I've written the reply too fast and with few details. First, we should add explicit triplets to fastify the plugins we maintain to make Node.js' ESM context happy while leaving the TS exports as ESM triplets to make TS happy.
Then we should patch FP package itself. Am I missing something?

**Comment by climba03003** [comment]
> First, we should add explicit triplets to fastify the plugins we maintain to make Node.js' ESM context happy while leaving the TS exports as ESM triplets to make TS happy.
Then we should patch FP package itself. Am I missing something?

Nothing missing.

For a complete support for all combination of both `NodeJS` and `TypeScript` in `ESM` and `CJS`.
The format of plugin and typings should be updated as follow.

```ts
// index.js

function plugin() {}

// CJS export
module.exports = plugin
// TypeScript ESM default export, not confirmed
module.exports.default = plugin
// CJS and ESM named export
module.exports.plugin = plugin

// index.d.ts

// type for plugin
type Plugin = FastifyPluginCallback<plugin.PluginOptions>

// required for CJS
namespace plugin {
  // other export of typings
  export interface PluginOptions {}

  export const plugin: Plugin
  // ESM default export
  export { default as plugin }
}

// actual plugin usage
declare function plugin(...params: Parameters<Plugin>): : ReturnType<Plugin>

// CJS module.exports
export = plugin


```

**Comment by climba03003** [comment]
> 
> Exporting namespace will break `"moduleInterop": false` compilation environments and force them to use import star for "default" imports.

No, it is supported by `module.exports.default`.
https://www.typescriptlang.org/tsconfig#esModuleInterop
You can check the second point of `esModuleInterop`

**Comment by fox1t** [comment]
I am talking about the types here (so what the compiler sees thanks to the type declarations), not the runtime (so `module.exports.default`). I would like to see a repro that pass all of the cases we support usual, when using `export =`. :/

**Comment by climba03003** [comment]
> I am talking about the types here (so what the compiler sees thanks to the type declarations), not the runtime (so `module.exports.default`). I would like to see a repro that passes all of the cases we support usual when using `export =`. :/

Yes, it is supported. You can use `@fastify/cookie` as an example.

```ts
// tsconfig.json
{
  "compilerOptions": {
    "esModuleInterop": false
  }
}

// index.ts
import FastifyCookie from '@fastify/cookie';
import Fastify from 'fastify';

const fastify = Fastify()
fastify.register(FastifyCookie)

;(async function() {
  await fastify.ready()
})()
```

**Comment by fox1t** [comment]
@climba03003 awesome. We should write up somewhere all of these findings to remember them in the future!

**Comment by fox1t** [comment]
@climba03003 In order to make it work, I needed to slightly adjust your types:
```
type Plugin = FastifyPluginCallback<plugin.PluginOptions>

// required for CJS
declare namespace plugin { // <- added declare
  // other export of typings
  export interface PluginOptions {}

  // CJS and ESM named export
  export const plugin: Plugin
  // ESM default export
  export { plugin as default } // <- instead of export { default as plugin } 
}

// actual plugin usage
declare function plugin(...params: Parameters<Plugin>): ReturnType<Plugin>

// CJS module.exports
export = plugin
```

**Comment by fox1t** [comment]
This is the explanation of what happened here (and in other Fastify packages) and why we need to start using the above types in the future!

https://github.com/microsoft/TypeScript/issues/50690#issuecomment-1241359555

**Comment by climba03003** [comment]
> This is the explanation of what happened here (and in other Fastify packages) and why we need to start using the above types in the future!
> 
> [microsoft/TypeScript#50690 (comment)](https://github.com/microsoft/TypeScript/issues/50690#issuecomment-1241359555)

Issue has been open in https://github.com/fastify/fastify/issues/4349

**Comment by Uzlopak** [comment]
Is this still relevant afterr #224 ?

**Comment by climba03003** [comment]
> Is this still relevant afterr #224 ?

Yes

**Pull request merged**